Change when http verification happens

td-extreme opened the pull request:
Fixed problem of HttpRequest object being able to be created from an invalid HttpRequest.
The ClientConnection now returns a String[] instead of a String.
The HttpServerRunner no longer creates the HttpRequest object.
The RequestHandler uses the validator to check if the request is valid. 
The RequestValidator now takes in a String[] and checks that the RequestLine is valid.

Currently the RequestHandler just returns HTTP/1.1 200 OK if the request is valid. 
In the future it will instead create an HttpRequest object and send that object to the proper request type handler.

td-extreme committed:
Changed when an HttpRequest is validated.

Verifier now takes a String[] instead of an HttpRequest object.
Verification is now performed before a HttpRequest object is created.
The HttpServerRunner no longer creates the HttpRequest object, it will
be created by the RequestHandler class in the future.

td-extreme committed:
Changed responses to valid HTTP Responses.

Also updated validation to handle wrong number of items in request
array.

mikeknep commented on src/main/java/RequestHandler.java:
```diff
@@ -2,17 +2,17 @@
+      return "HTTP/1.1 200 OK\r\n\r\nValid Http Request";
     } else {
-      return "400";
+      return "HTTP/1.0 400 Bad Request";
```
Why `1.0` and not `1.1`?

td-extreme replied:
Not sure how that happened.

mikeknep commented on src/main/java/ClientConnection.java:
```diff
@@ -16,7 +16,7 @@ public ClientConnection(int portNumber) throws IOException {
   }
 
-  public String receiveRequest() throws IOException {
+  public String[] receiveRequest() throws IOException {
```
I'm a little concerned about this return type, for a couple reasons:

1) Your `ClientConnection` is currently not tested. To what degree it should be tested, if at all, is debatable, but if it does remain not tested at all, or minimally tested, it should have as little logic around it as possible. IMO the `split` on whitespace below qualifies as enough additional logic that it should definitely be tested.

2) The `String[]` return type and splitting on whitespace demonstrates that knowledge about the request format is leaking into this `ClientConnection` class and away from some other object that should be responsible for understanding what constitutes an HTTP request. To put that another way: when this returned a `String`, it's only responsibility was to collect data from an input stream; it was something else's responsibility to understand HTTP and attempt to coerce that data into an HTTP request. Now we have this object splitting apart components of the HTTP request _and_ passing those components via array to a second object that now implicitly depends on this first object (to put the pieces of the request in the proper order).

3) `String[]` introduces the need to do an annoying check for `IndexOutOfBoundsException`. Blegh.

td-extreme replied:
I had made a StringParser class at first but then deleted it as it only had the one method that performed one line of code, the split(). I went ahead and rewrote that class and changed ConnectionClient back to the way it was, and updated the HttpServerRunner to use this new class.

td-extreme committed:
Created seperate Parser class.

td-extreme committed:
Fixed HTTP version number

dadarek commented on src/main/java/HttpServerRunner.java:
```diff
@@ -5,21 +5,23 @@
 
   public void runServer() throws IOException {
     String request;
+    String request_array[];
```
`camelCase`

dadarek commented on src/main/java/HttpVerifier.java:
```diff
@@ -1,12 +1,16 @@
+        isPathValid(request[1]) &&
+        isVersionValid(request[2]);
+    }
+    catch (IndexOutOfBoundsException e) {
```
While this works, it's not ideal to catch an `IndexOutOfBoundsException` if it can easily be checked for.

This also leads to a potentially misleading bug - imagine if your `isMethodValid` tries to incorrectly access some array that it creates throws an IndexOutOfBoundsException and doesn't catch it? Not a fun time debugging that.

dadarek commented on src/main/java/ParseString.java:
```diff
@@ -0,0 +1,8 @@
+public class ParseString implements IParser {
+
+  public String[] parse(String request) {
+    return request.split("\\s+");
```
Will this accept new lines and tabs as delimiters?

td-extreme replied:
Yes, it splits on space, /n, /r, and /t.

td-extreme committed:
Refactored HttpVerifier class.

Changed it so that instead of doing a bunch of or checks on a series of
boolean methods, that it instead uses a List<String> that contains all
the valid method types. It then checks if that List contains the method
being checked.

td-extreme committed:
Updated Test to check for /n/r/t

dadarek commented:
🚢  it

dadarek commented:
Could probably ship this so you don't have a PR hanging out for two weeks.
